stdlib/obj.ml

```diff
@@ -90,24 +90,67 @@ module Ephemeron = struct
 
   type t (** ephemeron *)
 
-  external create: int -> t = "caml_ephe_create"
+  let additional_values = 2
+  let max_ephe_length = Sys.max_array_length - additional_values
 
-  let length x = size(repr x) - 2
+  external create : int -> t = "caml_ephe_create";;
+  let create l =
+    if not (0 <= l && l <= max_ephe_length) then
+      invalid_arg "Obj.Ephemeron.create";
+    create l
+
+  let length x = size(repr x) - additional_values
+
+  let raise_if_invalid_offset e o msg =
+    if not (0 <= o && o < length e) then
+      invalid_arg msg
 
   external get_key: t -> int -> obj_t option = "caml_ephe_get_key"
+  let get_key e o =
+    raise_if_invalid_offset e o "Obj.Ephemeron.get_key";
+    get_key e o
+
   external get_key_copy: t -> int -> obj_t option = "caml_ephe_get_key_copy"
-  external set_key: t -> int -> obj_t -> unit = "caml_ephe_set_key"
-  external unset_key: t -> int -> unit = "caml_ephe_unset_key"
-  external check_key: t -> int -> bool = "caml_ephe_check_key"
+  let get_key_copy e o =
+    raise_if_invalid_offset e o "Obj.Ephemeron.get_key_copy";
+    get_key_copy e o
+
+  external set_key: t -> int -> obj_t -> unit = "caml_ephe_set_key" [@@noalloc]
+  let set_key e o x =
+    raise_if_invalid_offset e o "Obj.Ephemeron.set_key";
+    set_key e o x
+
+  external unset_key: t -> int -> unit = "caml_ephe_unset_key" [@@noalloc]
+  let unset_key e o =
+    raise_if_invalid_offset e o "Obj.Ephemeron.unset_key";
+    unset_key e o
+
+  external check_key: t -> int -> bool = "caml_ephe_check_key" [@@noalloc]
+  let check_key e o =
+    raise_if_invalid_offset e o "Obj.Ephemeron.check_key";
+    check_key e o
+
   external blit_key : t -> int -> t -> int -> int -> unit
-    = "caml_ephe_blit_key"
+    = "caml_ephe_blit_key" [@@noalloc]
+  let blit_key e1 o1 e2 o2 l =
+    let msg = "Obj.Ephemeron.blit_key" in
+    if l < 0 then invalid_arg msg;
+    if l = 0 then ()
+    else begin
+      raise_if_invalid_offset e1 o1 msg;
+      raise_if_invalid_offset e2 o2 msg;
+      let len1 = length e1 in let len2 = length e2 in
+      if not (l <= len1 && l <= len2 &&
+              o1 <= len1 - l && o2 <= len2 - l) then
+        invalid_arg msg;
```

Comment:
I think the above conditions are not sufficient to do the bounds checking.  Maybe this should be done once (in the C code instead of assert?)

Reply:
The `noalloc` annotations are in conflict with the design of ephemerons under multicore. Under multicore, we plan to root ephemerons to domains so that marking and sweeping of ephemerons are done by that domain. In order to avoid the race between the GC thread which sweeps the ephemeron and the mutator which may concurrently update the ephemeron, in the ephemeron sweeping phase, we require that the mutator requests the owning domain to perform the corresponding operation on the ephemeron. While the request is being serviced, the interruptor (i.e, the mutator) is expected to service other requests including a major GC cycling request. Hence, `noalloc` annotation is unsafe. The design of ephemerons under Multicore GC is being tracked in https://github.com/ocamllabs/ocaml-multicore/issues/88#issuecomment-382706882.

Reply:
I think it would be a mistake to add `noalloc` especially if we are adding a C API. The C code reasonably assumes that the GC would no run at calls to functions annotated `noalloc` and might choose to not register GC roots across such calls. In the future if we remove `noalloc` annotations, we will break code silently. 

> In which domain is an external C code?

I don't quite understand this question. Here is the answer to the question I think you are asking. External C code runs in the domain which invoked the C call. We will have a C api for creating additional domains and the C calls will run in that domain.

Reply:
I am not positive it should be addressed, but the following expression raises:
```
let a = Array.make 0 "" in Array.blit a 0 a 1 0
```
while the following one does not:
```
let w = Weak.create 0 in Weak.blit w 0 w 1 0
```
(because in `blit_key` source/destination offsets are checked iff the length
is strictly positive)
Is it what you had in mind @mshinwell?

Reply:
I don't have other concerns wrt multicore.

Reply:
I think the check is now correct.